include/SFML/Graphics/Rect.hpp

```diff
@@ -116,23 +117,7 @@ class Rect
     /// \see contains
     ///
     ////////////////////////////////////////////////////////////
-    constexpr bool intersects(const Rect<T>& rectangle) const;
-
-    ////////////////////////////////////////////////////////////
-    /// \brief Check the intersection between two rectangles
-    ///
-    /// This overload returns the overlapped rectangle in the
-    /// \a intersection parameter.
-    ///
-    /// \param rectangle    Rectangle to test
-    /// \param intersection Rectangle to be filled with the intersection
-    ///
-    /// \return True if rectangles overlap, false otherwise
-    ///
-    /// \see contains
-    ///
-    ////////////////////////////////////////////////////////////
-    constexpr bool intersects(const Rect<T>& rectangle, Rect<T>& intersection) const;
+    constexpr std::optional<Rect<T>> intersects(const Rect<T>& rectangle) const;
```

Comment:
The semantics haven't really changed so I think the name can stay. You can still use this as a boolean check if you don't care about the intersection rectangle.

Reply:
The semantics have changed: before, the function returned _whether_ there was an intersection, now it returns _the intersection itself_.

I was mostly thinking about this usage:
```cpp
if (auto intersection = rect.intersect(other))
   doSomethingWith(intersection);
else
   doNothing();
```

You're raising a good point though, that it's also possible to simply write ```if (rect.intersects(other))```.

I'd still say functions with 3rd person singular verb are almost always used for boolean checks, e.g. `isAdjective()`, `hasAttribute()`, `supportsCapability()`. So the current name might suggest that this is the only purpose of this function.

Reply:
The function still does everything it previously did. The same code is being called. It's still returning a truthy value. Almost nothing observable about it has changed. There's just this additional value that is also being returned alongside the truthy value that has always been there. Ultimately I don't feel strongly about this but `intersects` is the best name I can think of currently.

Reply:
I think `intersects` verb feels like a yes-or-no question: "Does it intersect?", while a noun `intersection` would imply an affirmation + potential result: "If it does intersect, give me the intersection".

IMO I like
```cpp
if (auto collisionPoint = rect.intersection(otherRect))
```

Reply:
```cpp
if (rect.intersection(otherRect))
```
reads a bit oddly to me compared to
```cpp
if (rect.intersects(otherRect))
```
but I'll accept either.

Reply:
To be honest, I don't think `intersection` is a good identifier, either. There are many APIs which use nouns as getters, among others the standard library (`vector::size()`, `iterator::end()`, `regex::flags()`). However, SFML does not do that, we use `get*()`.

I would still not use `getIntersection`, because it's not a getter, and this name suggests there is no work to do -- unlike `computeIntersection`, which is rather verbose though.

Why I suggested `intersect`: 
* it is the most general and a good compromise between conciseness and "don't mislead people". I read it as a command "rect, intersect the other rect" 🙂 
* That the intersection is returned is not _implied_ by the name, but I would say it's standing to reason, and definitely more so than with `intersects`.
* We have a precedent for this convention: [`transform.combine(transform)`](https://www.sfml-dev.org/documentation/2.5.1/classsf_1_1Transform.php#acd978f60421a0f839bb9a8263e8877ff).